openssh/USAGE.md

```diff
@@ -0,0 +1,128 @@
+
+Be aware that configurations for sshd and for ssh can be different as long as you don't want to connect to localhost. For example can ssh be configured like a classical ssh client and sshd may support post-quantum algorithms only with no backwards compatibility at all.
+
+Note that as long as the backwards compatibility is maintained, the system can not be considered post-quantum safe for obvious reasons.
+
+### Key re-generation
+
+When the image is run normally (`docker run -it oqs-img -t oqs-ssh`) it invokes the script [key-regen.sh](key-regen.sh). It generates all host and identity keys that **do not exist** and **are necessary** according to the configuration files ([ssh_config](ssh_config) and [sshd_config](sshd_config)). Their necessity is determinded based on the following parameters:
+1. `IdentityFile` for **identity keys**: For every entry (there may be multiple) the corresponding key is generated.
+   - e.g. `IdentityFile ~/.ssh/id_ed25519` or
+   - `IdentityFile ~/.ssh/id_p256_dilithium2`
+2. `HostKeyAlgorithms` for **host keys**: For every algorithm listed a host key will be generated.
+   - e.g. `HostKeyAlgorithms ssh-p256-dilithium2,ssh-ed25519`
+
+Note that the `key-regen.sh` script is executed as the `root` user.
+
+As mentioned above those keys will only be generated if they don't yet exist. So even though the script is executed every time a container is started (or run), it usually only does something the first time. If any host key was generated, the `sshd` service will be restarted.
+
+The location where `key-regen.sh` is looking for `ssh_config`/`sshd_config` is the install directory of `oqs-ssh`. The [Dockerfile](Dockerfile) puts this location into the variable `OQS_INSTALL_DIR` where it will be accessible from the script.
+## Using oqs-ssh for quantum-safe remote access with minimal intrusion
+
+One use case of quantum-safe ssh running in docker could be accessing a remote system without messing with its ssh(d) installation or other parts of the system you maybe don't want to interact with. This means minimal intrusion and everything can easily be removed again. This is done by running this docker image on said system and sharing its network space. Thus it is possible to access host ports from **within** the docker container. Normally, the use case of docker is one of isolation with some shared directories and published ports at max. So this solution works around the usual docker limitations.
+
+Additionally, it is advised to **change the default username and password** when building the image because your plan is to expose it to the world.
```

Comment:
Is there a way to force a deployer to explicitly pick a new password rather than suggesting it?

Reply:
I am not aware of a way to do this, and a really quick googlin' didn't bring any revelation. 

This actually brings up another topic I must admit I am not really happy with. As I have discussed with @baentsch some time ago, it may make more sense to switch to the oqs user in the Dockerfile and never start the container as root. But this makes then running the `sshd` a little more complicated. There are a few topics in the wild internet talking about running `sshd` without root privileges, but some of those solutions are deprecated and it does not seem common. I have tried some things, but nothing really worked to my satisfaction. And also from a security standpoint I it seems it is even recommended practice to run the `sshd` as root. So I sticked to using root. That basically means the oqs user is only used for remote login and not for local container management, which I can live with.

Reply:
I'm not a docker expert so I'm not sure what's the best course of action here. Following recommended practice is good, so I'm happy with it if we think it's the best ( slash not-overly-complicated ) option.